# **1. 서울 열린데이터 광장**
[서울 열린데이터 광장]('https://data.seoul.go.kr/')(Seoul Open Data Plaza)은 서울시에서 운영하는 공공데이터 개방 플랫폼입니다. 시민, 연구자, 기업 등이 서울시에서 생성한 다양한 공공데이터를 자유롭게 활용할 수 있도록 제공하고 있습니다. 이를 통해 데이터 기반의 창의적인 아이디어와 혁신을 촉진하며, 시민들의 정보 접근성을 높이고 공공서비스를 개선하는 데 기여하고 있습니다.

인증키 정보:62486256776c6f74353072744a4779

# **2. JSON**
JSON(JavaScript Object Notation)은 자바스크립트 객체 표기법을 기반으로 한 데이터 교환 형식으로, 사람도 읽기 쉽고 기계도 쉽게 분석하고 생성할 수 있는 텍스트 기반의 데이터 포맷입니다. 일반적으로 서버와 클라이언트 간에 데이터를 주고받을 때 많이 사용되며, 구조는 키-값 쌍으로 이루어진 객체 형태나 배열 형태를 사용합니다. 언어에 독립적이면서도 자바스크립트와 호환성이 뛰어나 웹 개발뿐만 아니라 다양한 API 통신, 데이터 저장 및 전송 등에 널리 활용됩니다.

### JSON 문법 구조

1. 객체(Object)
- 중괄호 {} 사용
- 내부에 "key": value 형식으로 속성-값 쌍을 나열
- 각 쌍은 쉼표 ,로 구분
- 값으로는 문자열, 숫자, 불리언, 배열, 객체 모두 가능

```
{
  "name": "김사과",
  "age": 20,
  "isStudent": true,
  "skills": ["HTML", "CSS", "JavaScript"]
}
```

### 2. 배열(Array)
- 대괄호 [] 사용
- 여러 개의 값을 순서대로 나열
- 요소는 숫자, 문자열, 객체 등 다양한 데이터 가능

```
[
  "사과",
  "바나나",
  "오렌지"
]
```

### 3. JSON 안에 JSON (중첩 구조)
- 객체 안에 객체, 객체 안에 배열도 가능
- 실제 API 응답이나 설정 파일에서 많이 사용됨

```
{
  "name": "김사과",
  "address": {
    "city": "서울",
    "zipcode": "12345"
  },
  "skills": ["HTML", "CSS", "JavaScript"]
}
```

### 4. JSON 규칙 (중요)
- 문자열은 반드시 큰따옴표 " ", 작은따옴표 ' '는 에러 발생
- 키(key)도 반드시 쌍따옴표 " "로 감싸야 함
- 값(value)에는 객체, 배열, 문자열, 숫자, 불리언, null 사용 가능
- 마지막 항목 뒤에는 쉼표 금지
- JSON은 순수 데이터 형식이기 때문에 //, /* */ 같은 주석 사용 불가

# **2. 서울시 공공자전거 실시간 대여정보**

In [37]:
import requests  # 서울시 공공데이터 인터넷 서버에 접속하는 모듈
import folium  # 지도 그리는 데이터 모듈
import json   # json데이터를 다룰 수 있는 모듈
import pandas as pd   # 데이터프레임 만들때 쓰는 모듈
import warnings   # 필터 걸러주는 모듈.
warnings.filterwarnings('ignore')

In [38]:
base_ur1 = "http://openapi.seoul.go.kr:8088/62486256776c6f74353072744a4779/json/bikeList/1/5/"
response = requests.get(base_ur1)
response #200번때는 정상적인 수치이고 400번대 페이지는 호출을 못한 수치이다.

<Response [200]>

In [39]:
json_data = response.json()
json_data

{'rentBikeStatus': {'list_total_count': 5,
  'RESULT': {'CODE': 'INFO-000', 'MESSAGE': '정상 처리되었습니다.'},
  'row': [{'rackTotCnt': '15',
    'stationName': '102. 망원역 1번출구 앞',
    'parkingBikeTotCnt': '13',
    'shared': '87',
    'stationLatitude': '37.55564880',
    'stationLongitude': '126.91062927',
    'stationId': 'ST-4'},
   {'rackTotCnt': '14',
    'stationName': '103. 망원역 2번출구 앞',
    'parkingBikeTotCnt': '16',
    'shared': '114',
    'stationLatitude': '37.55495071',
    'stationLongitude': '126.91083527',
    'stationId': 'ST-5'},
   {'rackTotCnt': '13',
    'stationName': '104. 합정역 1번출구 앞',
    'parkingBikeTotCnt': '2',
    'shared': '15',
    'stationLatitude': '37.55073929',
    'stationLongitude': '126.91508484',
    'stationId': 'ST-6'},
   {'rackTotCnt': '5',
    'stationName': '105. 합정역 5번출구 앞',
    'parkingBikeTotCnt': '1',
    'shared': '20',
    'stationLatitude': '37.55000687',
    'stationLongitude': '126.91482544',
    'stationId': 'ST-7'},
   {'rackTotCnt': '12',


In [40]:
json_data.get("rentBikeStatus",{}).get("RESULT",{}).get("CODE","")

'INFO-000'

In [41]:
json_data["rentBikeStatus"]["RESULT"]["CODE"]

'INFO-000'

두 표현의 **차이점은 “키가 없을 때 에러가 나느냐, 빈 값을 주느냐”**야.

```
✅ 1. json_data.get("rentBikeStatus",{}).get("RESULT",{}).get("CODE","")
✔️ 특징: 에러 안 나고 안전함 (Safe 접근)

.get(key, 기본값) 방식은 해당 key가 없으면 에러 대신 기본값을 반환해.

기본값을 {} 또는 "" 로 지정했으므로, 중간에 어느 키가 없어도 프로그램이 멈추지 않아.

예시
json_data = {}   # rentBikeStatus 자체가 없음

value = json_data.get("rentBikeStatus", {}).get("RESULT", {}).get("CODE", "")
print(value)  # ""   ← 에러 없이 빈 문자열 출력
```
```
✅ 2. json_data["rentBikeStatus"]["RESULT"]["CODE"]
✔️ 특징: 키가 없으면 바로 오류 발생 (KeyError)

json_data["rentBikeStatus"]

"rentBikeStatus"가 없으면 KeyError

"RESULT"가 없으면 KeyError

"CODE"가 없으면 KeyError

즉, 구조가 100% 보장된 JSON일 때만 안전하게 쓸 수 있어.

```

| 표현                       | 안전성      | 키 없을 때      | 언제 사용?             |
| ------------------------ | -------- | ----------- | ------------------ |
| `dict.get().get().get()` | 👍 매우 안전 | 기본값 반환      | 데이터가 불완전할 수 있을 때   |
| `dict["key"]["key"]`     | ❌ 위험     | KeyError 발생 | 반드시 키가 존재한다고 확신할 때 |



In [49]:
import requests
import pandas as pd

def fetch_bike_data():
    base_url = "http://openapi.seoul.go.kr:8088/62486256776c6f74353072744a4779/json/bikeList/"
    start = 1
    end = 1000
    step = 1000
    data_frames = []

    while True:
        # 예: "http://openapi.seoul.go.kr:8088/.../json/bikeList/1/1000"
        url = f"{base_url}{start}/{end}/"
        response = requests.get(url)

        if response.status_code != 200:
            print(f"Status Code: {response.status_code}")
            break

        json_data = response.json()

        try:
            rent_bike_status = json_data["rentBikeStatus"]
            result_code = rent_bike_status["RESULT"]["CODE"]
        except KeyError:
            print("JSON 오류")
            break

        if result_code == "INFO-200":
            print("데이터 없음")
            break

        elif result_code == "INFO-000":
            print(f"시작: {start} 끝: {end}")
            try:
                bike_data = rent_bike_status["row"]
                if bike_data:
                    df = pd.DataFrame(bike_data)
                    data_frames.append(df)
            except KeyError:
                print("데이터를 찾을 수 없음")

        elif result_code == "ERROR-336":
            print("데이터요청은 한번에 최대 1,000건을 넘을 수 없습니다")
            break

        else:
            print(f"result code: {result_code}")
            break

        start += step
        end += step

    # ✅ while 끝난 뒤에 한 번만 최종 반환
    if data_frames:
        final_df = pd.concat(data_frames, ignore_index=True)
        return final_df
    else:
        return pd.DataFrame()


In [50]:
bike_data_df = fetch_bike_data()

시작: 1 끝: 1000
시작: 1001 끝: 2000
시작: 2001 끝: 3000
JSON 오류


In [51]:
bike_data_df

rackTotCnt            stationName parkingBikeTotCnt shared  \
0            15        102. 망원역 1번출구 앞                13     87   
1            14        103. 망원역 2번출구 앞                16    114   
2            13        104. 합정역 1번출구 앞                 2     15   
3             5        105. 합정역 5번출구 앞                 1     20   
4            12        106. 합정역 7번출구 앞                 6     50   
...         ...                    ...               ...    ...   
2753          8             6185.가양나들목                11    138   
2754         12     6187.마곡119안전센터 맞은편                19    158   
2755         10             6188.금호아파트                 3     30   
2756         11     6189.데시앙플렉스 지식산업센터                16    145   
2757         10  6190.마곡광장(마곡나루역 5번출구)                 3     30   

     stationLatitude stationLongitude stationId  
0        37.55564880     126.91062927      ST-4  
1        37.55495071     126.91083527      ST-5  
2        37.55073929     126.91508484      ST-6  
3        37.55000687     126.91482544      ST-7  
4        37.54864502     126.91282654      ST-8  
...              ...              ...       ...  
2753     37.57341003     126.84345245   ST-3418  
2754     37.55534744     126.82072449   ST-3415  
2755     37.55619049     126.86463928   ST-3419  
2756     37.56448364     126.84830475   ST-3424  
2757     37.56614685     126.82738495   ST-3421  

[2758 rows x 7 columns]

In [52]:
bike_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2758 entries, 0 to 2757
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   rackTotCnt         2758 non-null   object
 1   stationName        2758 non-null   object
 2   parkingBikeTotCnt  2758 non-null   object
 3   shared             2758 non-null   object
 4   stationLatitude    2758 non-null   object
 5   stationLongitude   2758 non-null   object
 6   stationId          2758 non-null   object
dtypes: object(7)
memory usage: 151.0+ KB


In [53]:
'''
rackTotCnt	거치대개수
parkingBikeTotCnt	자전거주차총건수
shared	거치율
stationLatitude	위도
stationLongitude	경도
stationId	대여소ID
stationName	대여소이름
'''
bike_data_df.columns

Index(['rackTotCnt', 'stationName', 'parkingBikeTotCnt', 'shared',
       'stationLatitude', 'stationLongitude', 'stationId'],
      dtype='object')

In [54]:
bike_data_df['stationLongitude'] = bike_data_df['stationLongitude'].astype(float)
bike_data_df['stationLatitude'] = bike_data_df['stationLatitude'].astype(float)

In [55]:
bike_map = folium.Map(location=[bike_data_df['stationLatitude'].mean(),
                                bike_data_df['stationLongitude'].mean()],
                                zoom_start=12)

for index, data in bike_data_df.iterrows():
    popup_str = '{} 자전거주차총건수:{}대'.format(
        data['stationName'], data['parkingBikeTotCnt']
    )
    popup = folium.Popup(popup_str, max_width=600)
    folium.Marker(location=[data['stationLatitude'], data['stationLongitude']],
                  popup=popup).add_to(bike_map)

bike_map

### 1. 지도 객체 만들기
bike_map = folium.Map(location=[bike_data_df['stationLatitude'].mean(),
                                bike_data_df['stationLongitude'].mean()],
                                zoom_start=12)

🔹 folium.Map(...)

Folium에서 지도의 “바탕”을 만드는 함수야.
이 bike_map 위에 마커를 하나씩 올려 붙이는 구조라고 보면 돼.

🔹 location=[ ... , ... ]

지도에서 처음 화면의 중심 좌표 (위도, 경도) 를 의미해.

여기서는:

bike_data_df['stationLatitude'].mean()
bike_data_df['stationLongitude'].mean()


이렇게 전체 대여소 위도/경도의 평균값을 사용하고 있어.

👉 즉,

모든 자전거 대여소의 위치를 쭉 모아서

그 중심 근처를 지도 가운데로 잡는 셈이야.
서울시 공공자전거 전체를 보는 거라면 합리적인 선택이야.

전제: stationLatitude, stationLongitude 컬럼이 숫자(float)여야 mean()이 잘 동작해.

🔹 zoom_start=12

처음에 지도를 얼마나 확대해서 보여줄지를 결정해.

숫자가 클수록 더 확대됨.

10~12 : 도시 전체

14~16 : 동네 단위

12면 “서울시 전체를 대충 한 화면에 보는” 느낌이라고 생각하면 돼.

### 2. 데이터프레임을 한 줄씩 돌면서 마커 찍기
for index, data in bike_data_df.iterrows():

🔹 bike_data_df.iterrows()

bike_data_df의 각 행(row)을 하나씩 꺼내는 반복문이야.

한 번 돌 때마다:

index → 그 행의 인덱스 번호 (0, 1, 2, …)

data → 그 행의 내용이 담긴 Series 객체

예를 들어 한 행이 이렇게 생겼다고 하면:

stationName	stationLatitude	stationLongitude	parkingBikeTotCnt
서울시청 앞	37.56...	126.97...	15

반복문 안에서는

data['stationName']            → "서울시청 앞"
data['stationLatitude']        → 37.56...
data['stationLongitude']       → 126.97...
data['parkingBikeTotCnt']      → 15


이런 식으로 접근할 수 있게 되는 거야.

### 3. 팝업에 보여줄 문자열 만들기
    popup_str = '{} 자전거주차총건수:{}대'.format(
        data['stationName'], data['parkingBikeTotCnt']
    )

🔹 str.format(...) 사용

{} 부분이 순서대로 .format() 안의 값으로 채워져.

예를 들어

data['stationName'] = "서울시청 앞"

data['parkingBikeTotCnt'] = 15

그러면 popup_str은:

"서울시청 앞 자전거주차총건수:15대"


이렇게 되는 거지.

👉 나중에 마커를 클릭했을 때 말풍선으로 뜨는 내용이 이 문자열이야.

### 4. 팝업 객체 만들기
    popup = folium.Popup(popup_str, max_width=600)

🔹 folium.Popup(...)

방금 만든 popup_str을 팝업(말풍선) 객체로 래핑한 거야.

max_width=600은 팝업이 너무 길어졌을 때 최대 가로 길이를 제한해 주는 옵션.

줄이 너무 길면 보기 불편하니까, 대략 600px 넘지 않게 조절.

이 popup은 잠시 뒤에 마커에 붙여줄 거야.


### 5. 지도에 마커(핀) 하나 추가하기
    folium.Marker(location=[data['stationLatitude'], data['stationLongitude']],
                  popup=popup).add_to(bike_map)

🔹 folium.Marker(...)

지도 위에 찍는 기본 핀(마커) 야.

location=[위도, 경도]
→ 이 마커가 찍힐 위치.

여기선:

location=[
    data['stationLatitude'],
    data['stationLongitude']
]


이니까,
각 대여소의 좌표에 맞춰 마커가 하나씩 찍히는 구조야.

🔹 popup=popup

앞에서 만든 folium.Popup 객체를 이 마커에 연결해주는 것.

실제 동작:

사용자가 마커를 클릭하면

popup_str으로 만든 텍스트가 말풍선으로 뜸
→ "역이름 자전거주차총건수:OO대" 형식

🔹 .add_to(bike_map)

이 마커를 어떤 지도에 올릴지를 지정하는 부분.

bike_map은 제일 처음에 만든 지도 객체지?

즉, “이 마커를 bike_map이라는 지도에 추가해라” 라는 뜻.

반복문이 도는 동안
→ 대여소 개수만큼 마커가 계속 bike_map 위에 쌓이는 셈이야.

### 6. 마지막 줄: 지도 객체 보여주기
bike_map


Jupyter / Colab / IPython 환경에서는
마지막 줄에 객체 하나만 두면 그걸 화면에 표시해 줘.

Folium의 Map 객체는
→ 내부적으로 HTML + JavaScript(Leaflet)를 만들어서
→ 셀 출력 영역에 지도를 렌더링해.

그래서 이 한 줄이 “지도를 출력해라”라는 의미가 되는 거야.

### 7. 전체 코드 흐름 한 번에 요약하면

위도/경도 평균으로 지도 중심을 잡고 bike_map 생성

bike_data_df의 각 대여소 행을 하나씩 돌면서

역 이름 + 자전거 수로 팝업 문자열 만들고

그걸 folium.Popup으로 감싸고

해당 좌표에 folium.Marker를 생성해서

팝업을 붙인 뒤 bike_map에 추가

마지막에 bike_map을 출력해서
→ “서울시의 모든 대여소 위치 + 자전거 대수 정보가 담긴 지도”가 만들어짐